[AutoDiff] SILGen differentiability witnesses.

dan-zheng opened the pull request:
Generate SIL differentiability witnesses from AST `@differentiable` attributes,
using lowered parameter indices, result indices (currently with capacity 1 and
set index 0), and derivative generic signature.

Resolves [TF-869](https://bugs.swift.org/browse/TF-869).

---

The [TF-866](https://bugs.swift.org/browse/TF-866) master issue tracks all retroactive derivative registration tasks.

Todos related to this patch:
- [TF-835](https://bugs.swift.org/browse/TF-835): lower `@differentiating` attribute to SIL `[differentiable]` attribute instead of creating implicit `@differentiable` attributes. This is blocked by full retroactive derivative registration, see issue comment for details.
- [TF-919](https://bugs.swift.org/browse/TF-919): explore emitting serialized differentiability witnesses. Currently, SILGen never creates serialized differentiability witnesses to avoid deserialization issues.

dan-zheng commented on test/AutoDiff/sil_differentiability_witness_silgen.swift:
```diff
@@ -0,0 +1,54 @@
+// RUN: %target-swift-frontend -emit-silgen %s | %target-sil-opt | %FileCheck %s
+
+// Test SIL differentiability witness SIL generation.
```
Todo: add more tests, namely `@differentiable`/`@differentiating` attributes for different original declaration kinds (initializers, subscripts, methods, computed properties).

dan-zheng replied:
Added some more tests.

dan-zheng committed:
[AutoDiff] SILGen differentiability witnesses.

Generate SIL differentiability witnesses from AST `@differentiable` attributes,
using lowered parameter indices, result indices (currently with capacity 1 and
set index 0), and derivative generic signature.

Resolves TF-869.

The TF-866 master issue tracks all retroactive derivative registration tasks.

dan-zheng commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3057,6 +3057,51 @@ void SILDefaultWitnessTable::dump() const {
 
+// TODO(TF-893): Use this helper to dedupe the same logic in
+// `SILFunction::print`.
+static void printSILFunctionNameAndType(
```
Note: these printer changes were necessary to fix round-trip printing/parsing errors for generic signature requirements:
```
<stdin>:1864:88: error: use of undeclared type 'T'
sil_differentiability_witness hidden [parameters 0 1] [results 0]  [where T : _Differentiable]  @generic : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0 {
```

Copying logic from `SILFunction::print` was the only way I found to fix the errors (by printing the generic signature requirements in the original function's generic environment).

dan-zheng replied:
Actually, I found out how to print `τ_0_0 `: https://gist.github.com/dan-zheng/96661bc958f05374a2615f2b7aac5867.

Will keep printing `T` since it is the output of SILGen.

dan-zheng commented on lib/SILGen/SILGen.cpp:
```diff
@@ -752,87 +752,140 @@ void SILGenModule::postEmitFunction(SILDeclRef constant,
+
+  // Get or create differentiability witness.
+  CanGenericSignature derivativeGenSig;
+  if (jvp && vjp)
```
Noted, thanks. The code in question was changed so the suggestion no longer applies.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3097,26 +3142,32 @@ void SILDifferentiabilityWitness::print(
+  if (vjp) {
+    OS << "  vjp: ";
+    printSILFunctionNameAndType(OS, vjp);
+    OS << "\n";
```
Use character literal.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27652/commits/fa71b2609d18a32ac158ae7b9e7e7f0dc013df90.

rxwei replied:
Same here.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3097,26 +3142,32 @@ void SILDifferentiabilityWitness::print(
+  if (jvp) {
+    OS << "  jvp: ";
+    printSILFunctionNameAndType(OS, jvp);
+    OS << "\n";
```
Use character literal.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27652/commits/fa71b2609d18a32ac158ae7b9e7e7f0dc013df90.

rxwei replied:
It seems not done.

dan-zheng replied:
Sorry, the change got lost in an accidental revert. I'll fix in a follow-up.

rxwei commented on lib/SILGen/SILGen.h:
```diff
@@ -318,6 +318,17 @@ class LLVM_LIBRARY_VISIBILITY SILGenModule : public ASTVisitor<SILGenModule> {
+  /// functions (null if undefined).
+  void emitDifferentiabilityWitness(AbstractFunctionDecl *originalAFD,
+                                    SILFunction *originalFunction,
+                                    IndexSubset *parameterIndices,
```
Use a `const SILAutoDiffConfig &` parameter instead.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27652/commits/fa71b2609d18a32ac158ae7b9e7e7f0dc013df90.

rxwei commented on test/AutoDiff/sil_differentiability_witness.sil:
```diff
@@ -74,7 +74,7 @@ sil_differentiability_witness hidden [parameters 0 1] [results 0] [where τ_0_0
 
 // CHECK-LABEL: // differentiability witness for generic
-// CHECK: sil_differentiability_witness hidden [parameters 0 1] [results 0] [where τ_0_0 : _Differentiable] @generic : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0 {
+// CHECK: sil_differentiability_witness hidden [parameters 0 1] [results 0] [where T : _Differentiable] @generic : $@convention(thin) <T> (@in_guaranteed T, Float) -> @out T {
```
`τ_0_0` feels better than `T` because 1) it is what is parsed and 2) SILFunction's use `τ_0_0` when they come out of SILGen.

dan-zheng replied:
Actually, SILGen preserves `T`, so let's preserve it.
```swift
func foo<T>(_ x: T) {}
```

```console
$ swiftc -emit-silgen foo.swift
...
// foo<A>(_:)
sil hidden [ossa] @$s3fooAAyyxlF : $@convention(thin) <T> (@in_guaranteed T) -> () {
// %0                                             // user: %1
bb0(%0 : $*T):
  debug_value_addr %0 : $*T, let, name "x", argno 1 // id: %1
  %2 = tuple ()                                   // user: %3
  return %2 : $()                                 // id: %3
} // end sil function '$s3fooAAyyxlF'
```

[This comment](https://github.com/apple/swift/pull/27652/files#r335202642) has some more context.

dan-zheng committed:
Address review feedback.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
CI failed:
```
swift: /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/llvm/include/llvm/ADT/SmallVector.h:155: llvm::SmallVectorTemplateCommon::reference llvm::SmallVectorTemplateCommon<unsigned int, void>::operator[](llvm::SmallVectorTemplateCommon::size_type) [T = unsigned int]: Assertion `idx < size()' failed.
Stack dump:
0.	Program arguments: /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/Ninja-ReleaseAssert+stdlib-Release/swift-linux-x86_64/bin/swift -frontend -emit-module -filelist /tmp/sources-0fc1a7 -supplementary-output-file-map /tmp/supplementaryOutputs-7871dc -disable-objc-attr-requires-foundation-module -target x86_64-unknown-linux-gnu -disable-objc-interop -sdk / -I /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/Ninja-ReleaseAssert+stdlib-Release/swift-linux-x86_64/./lib/swift/linux/x86_64 -warn-swift3-objc-inference-complete -warn-implicit-overrides -enable-library-evolution -module-cache-path /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/Ninja-ReleaseAssert+stdlib-Release/swift-linux-x86_64/./module-cache -module-link-name swiftCore -nostdimport -parse-stdlib -resource-dir /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/Ninja-ReleaseAssert+stdlib-Release/swift-linux-x86_64/./lib/swift -swift-version 5 -O -enforce-exclusivity=unchecked -group-info-path /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/swift/stdlib/public/core/GroupInfo.json -verify-syntax-tree -Xllvm -sil-inline-generics -Xllvm -sil-partial-specialization -Xcc -DswiftCore_EXPORTS -parse-as-library -module-name Swift -o /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/Ninja-ReleaseAssert+stdlib-Release/swift-linux-x86_64/./lib/swift/linux/x86_64/Swift.swiftmodule -runtime-compatibility-version none -disable-autolinking-runtime-compatibility-dynamic-replacements 
1.	Swift version 5.1-dev (LLVM c5340df2d1, Swift 11dcca990d)
2.	Contents of /tmp/sources-0fc1a7:
3.	While emitting SIL for 'differentiableFunction(from:)' (at /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/swift/stdlib/public/core/DifferentiationSupport.swift:214:8)
4.	While silgen emitFunction SIL function "@$ss22differentiableFunction4fromq_xXGq_5value_13TangentVectorQzADQy_c8pullbacktxc_ts15_DifferentiableRzsAHR_r0_lF".
 for 'differentiableFunction(from:)' (at /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/swift/stdlib/public/core/DifferentiationSupport.swift:214:8)
5.	While emitting SIL for 'original(_:)' (at /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/swift/stdlib/public/core/DifferentiationSupport.swift:218:3)
6.	While silgen emitFunction SIL function "@$ss22differentiableFunction4fromq_xXGq_5value_13TangentVectorQzADQy_c8pullbacktxc_ts15_DifferentiableRzsAHR_r0_lF8originalL_yq_xsAHRzsAHR_r0_lF".
 for 'original(_:)' (at /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/swift/stdlib/public/core/DifferentiationSupport.swift:218:3)
```

I didn't encounter this error until pulling from `apple/tensorflow`, investigating now.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
CI fails due to differentiability witness deserialization failures (see occurrences of `UNKNOWN`):
```
  Block ID #14 (SIL_BLOCK):
      Num Instances: 1
         Total Size: 12694b/1586.75B/396W
    Percent of file: 4.9487%
      Num SubBlocks: 0
        Num Abbrevs: 46
        Num Records: 56
    Percent Abbrevs: 100.0000%

	Record Histogram:
		  Count    # Bits     b/Rec   % Abv  Record Kind
		     16      3040     190.0  100.00  SIL_FUNCTION
		     10      2221     222.1  100.00  UnknownCode35
		      8       640      80.0  100.00  SIL_ONE_OPERAND
		      6      1144     190.7  100.00  SIL_INST_APPLY
		      4       452     113.0  100.00  SIL_ONE_TYPE_ONE_OPERAND
		      4       196      49.0  100.00  SIL_ONE_TYPE
		      4       792     198.0  100.00  SIL_BASIC_BLOCK
		      2       198      99.0  100.00  decls_block::PROTOCOL_CONFORMANCE_XREF
		      2       534     267.0  100.00  SIL_INST_WITNESS_METHOD

  Block ID #15 (SIL_INDEX_BLOCK):
      Num Instances: 1
         Total Size: 5686b/710.75B/177W
    Percent of file: 2.2167%
      Num SubBlocks: 0
        Num Abbrevs: 2
        Num Records: 4
    Percent Abbrevs: 100.0000%

	Record Histogram:
		  Count    # Bits     b/Rec   % Abv  Record Kind
		      1       324            100.00  SIL_DIFFERENTIABILITY_WITNESS_OFFSETS
		      1      1762            100.00  UnknownCode12
		      1       510            100.00  SIL_FUNC_OFFSETS
		      1      2936            100.00  SIL_FUNC_NAMES
```

These surfaced only after differentiability witness SILGen. Investigating now.

dan-zheng committed:
Fix serialization.

Fix serialization for Serialization/differentia{ble,ting}_attr.swift.

dan-zheng committed:
Always emit non-serialized differentiability witnesses.

Fix deserialization for Serialization/differentia{ble,ting}_attr.swift.
TF-919 has more info.

dan-zheng committed:
Clean up.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
